Error: string is required when updating from 1.10.0 to 1.10.1

`Error: string is required` with `terraform plan` when we update from `1.10.0` to `1.10.1`

We are still using Terraform v0.12.x

Currently, trying to find which string is required.

@rawmind0 Debug log
```
2020/09/01 14:11:50 [DEBUG] ReferenceTransformer: "rancher2_cluster.cluster" references: []
2020-09-01T14:11:50.252Z [DEBUG] plugin.terraform-provider-rancher2_v1.10.1: 2020/09/01 14:11:50 [WARN] unexpected type cty.String for map in json state
2020-09-01T14:11:50.252Z [DEBUG] plugin.terraform-provider-rancher2_v1.10.1: 2020/09/01 14:11:50 [WARN] unexpected type cty.String for map in json state
2020/09/01 14:11:50 [ERROR] <root>: eval: *terraform.EvalReadState, err: string is required
2020/09/01 14:11:50 [ERROR] <root>: eval: *terraform.EvalSequence, err: string is required
```

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -1,3 +1,17 @@
+## 1.10.3 (September 14, 2020)
+
+FEATURES:
+
+
+
+ENHANCEMENTS:
+
+
+
+BUG FIXES:
+
+* Fix `Error: string is required` upgrading rancher2 provider from v1.10.0 or lower
+
 ## 1.10.2 (September 10, 2020)
 
 FEATURES:

--- a/rancher2/data_source_rancher2_cluster.go
+++ b/rancher2/data_source_rancher2_cluster.go
@@ -28,7 +28,7 @@ func dataSourceRancher2Cluster() *schema.Resource {
 				MaxItems: 1,
 				Computed: true,
 				Elem: &schema.Resource{
-					Schema: clusterRKEConfigFields(),
+					Schema: clusterRKEConfigFieldsData(),
 				},
 			},
 			"k3s_config": {

--- a/rancher2/data_source_rancher2_cluster_template.go
+++ b/rancher2/data_source_rancher2_cluster_template.go
@@ -40,7 +40,7 @@ func dataSourceRancher2ClusterTemplate() *schema.Resource {
 				Computed:    true,
 				Description: "Cluster template revisions",
 				Elem: &schema.Resource{
-					Schema: clusterTemplateRevisionFields(),
+					Schema: clusterTemplateRevisionFieldsData(),
 				},
 			},
 			"annotations": {

--- a/rancher2/resource_rancher2_cluster.go
+++ b/rancher2/resource_rancher2_cluster.go
@@ -20,7 +20,15 @@ func resourceRancher2Cluster() *schema.Resource {
 		Importer: &schema.ResourceImporter{
 			State: resourceRancher2ClusterImport,
 		},
-		Schema: clusterFields(),
+		Schema:        clusterFields(),
+		SchemaVersion: 1,
+		StateUpgraders: []schema.StateUpgrader{
+			{
+				Type:    resourceRancher2ClusterResourceV0().CoreConfigSchema().ImpliedType(),
+				Upgrade: resourceRancher2ClusterStateUpgradeV0,
+				Version: 0,
+			},
+		},
 		// Setting default timeouts to be liberal in order to accommodate managed Kubernetes providers like EKS, GKE, and AKS
 		Timeouts: &schema.ResourceTimeout{
 			Create: schema.DefaultTimeout(30 * time.Minute),
@@ -28,6 +36,66 @@ func resourceRancher2Cluster() *schema.Resource {
 			Delete: schema.DefaultTimeout(30 * time.Minute),
 		},
 	}
+}
+
+func resourceRancher2ClusterResourceV0() *schema.Resource {
+	return &schema.Resource{
+		Schema: clusterFieldsV0(),
+	}
+}
+
+func resourceRancher2ClusterStateUpgradeV0(rawState map[string]interface{}, meta interface{}) (map[string]interface{}, error) {
+	if rkeConfigs, ok := rawState["rke_config"].([]interface{}); ok && len(rkeConfigs) > 0 {
+		for i1 := range rkeConfigs {
+			if rkeConfig, ok := rkeConfigs[i1].(map[string]interface{}); ok && len(rkeConfig) > 0 {
+				if services, ok := rkeConfig["services"].([]interface{}); ok && len(services) > 0 {
+					for i2 := range services {
+						if service, ok := services[i2].(map[string]interface{}); ok && len(service) > 0 {
+							if kubeApis, ok := service["kube_api"].([]interface{}); ok && len(kubeApis) > 0 {
+								for i3 := range kubeApis {
+									if kubeAPI, ok := kubeApis[i3].(map[string]interface{}); ok && len(kubeAPI) > 0 {
+										if eventRates, ok := kubeAPI["event_rate_limit"].([]interface{}); ok && len(eventRates) > 0 {
+											for i4 := range eventRates {
+												if eventRate, ok := eventRates[i4].(map[string]interface{}); ok && len(eventRate) > 0 {
+													if config, ok := eventRate["configuration"].(map[string]interface{}); ok {
+														newValue := ""
+														if len(config) > 0 {
+															conf, err := mapInterfaceToYAML(config)
+															if err == nil {
+																newValue = conf
+															}
+														}
+														rawState["rke_config"].([]interface{})[i1].(map[string]interface{})["services"].([]interface{})[i2].(map[string]interface{})["kube_api"].([]interface{})[i3].(map[string]interface{})["event_rate_limit"].([]interface{})[i4].(map[string]interface{})["configuration"] = newValue
+													}
+												}
+											}
+										}
+										if secretEncs, ok := kubeAPI["secrets_encryption_config"].([]interface{}); ok && len(secretEncs) > 0 {
+											for i4 := range secretEncs {
+												if secretEnc, ok := secretEncs[i4].(map[string]interface{}); ok && len(secretEnc) > 0 {
+													if config, ok := secretEnc["custom_config"].(map[string]interface{}); ok {
+														newValue := ""
+														if len(config) > 0 {
+															conf, err := mapInterfaceToYAML(config)
+															if err == nil {
+																newValue = conf
+															}
+														}
+														rawState["rke_config"].([]interface{})[i1].(map[string]interface{})["services"].([]interface{})[i2].(map[string]interface{})["kube_api"].([]interface{})[i3].(map[string]interface{})["secrets_encryption_config"].([]interface{})[i4].(map[string]interface{})["custom_config"] = newValue
+													}
+												}
+											}
+										}
+									}
+								}
+							}
+						}
+					}
+				}
+			}
+		}
+	}
+	return rawState, nil
 }
 
 func resourceRancher2ClusterCreate(d *schema.ResourceData, meta interface{}) error {

--- a/rancher2/resource_rancher2_cluster_template.go
+++ b/rancher2/resource_rancher2_cluster_template.go
@@ -19,7 +19,15 @@ func resourceRancher2ClusterTemplate() *schema.Resource {
 		Importer: &schema.ResourceImporter{
 			State: resourceRancher2ClusterTemplateImport,
 		},
-		Schema: clusterTemplateFields(),
+		Schema:        clusterTemplateFields(),
+		SchemaVersion: 1,
+		StateUpgraders: []schema.StateUpgrader{
+			{
+				Type:    resourceRancher2ClusterTemplateResourceV0().CoreConfigSchema().ImpliedType(),
+				Upgrade: resourceRancher2ClusterTemplateStateUpgradeV0,
+				Version: 0,
+			},
+		},
 		Timeouts: &schema.ResourceTimeout{
 			Create: schema.DefaultTimeout(10 * time.Minute),
 			Update: schema.DefaultTimeout(10 * time.Minute),
@@ -104,6 +112,33 @@ func resourceRancher2ClusterTemplate() *schema.Resource {
 			}),
 		),
 	}
+}
+
+func resourceRancher2ClusterTemplateResourceV0() *schema.Resource {
+	return &schema.Resource{
+		Schema: clusterTemplateFieldsV0(),
+	}
+}
+
+func resourceRancher2ClusterTemplateStateUpgradeV0(rawState map[string]interface{}, meta interface{}) (map[string]interface{}, error) {
+	if tmplRevisions, ok := rawState["template_revisions"].([]interface{}); ok && len(tmplRevisions) > 0 {
+		for i1 := range tmplRevisions {
+			if tmplRevision, ok := tmplRevisions[i1].(map[string]interface{}); ok && len(tmplRevision) > 0 {
+				if clusterConfigs, ok := tmplRevision["cluster_config"].([]interface{}); ok && len(clusterConfigs) > 0 {
+					for i2 := range clusterConfigs {
+						if clusterConfig, ok := clusterConfigs[i2].(map[string]interface{}); ok && len(clusterConfig) > 0 {
+							newValue, err := resourceRancher2ClusterStateUpgradeV0(clusterConfig, meta)
+							if err != nil {
+								return nil, fmt.Errorf("Upgrading Cluster Template schema V0: %v", err)
+							}
+							rawState["template_revisions"].([]interface{})[i1].(map[string]interface{})["cluster_config"].([]interface{})[i2] = newValue
+						}
+					}
+				}
+			}
+		}
+	}
+	return rawState, nil
 }
 
 func resourceRancher2ClusterTemplateCreate(d *schema.ResourceData, meta interface{}) error {

--- a/rancher2/schema_cluster.go
+++ b/rancher2/schema_cluster.go
@@ -94,6 +94,371 @@ func clusterAuthEndpoint() map[string]*schema.Schema {
 	return s
 }
 
+func clusterDataFieldsV0() map[string]*schema.Schema {
+	s := map[string]*schema.Schema{
+		"name": {
+			Type:     schema.TypeString,
+			Required: true,
+		},
+		"driver": {
+			Type:     schema.TypeString,
+			Computed: true,
+		},
+		"kube_config": {
+			Type:     schema.TypeString,
+			Computed: true,
+		},
+		"rke_config": {
+			Type:     schema.TypeList,
+			MaxItems: 1,
+			Optional: true,
+			Computed: true,
+			Elem: &schema.Resource{
+				Schema: clusterRKEConfigFieldsV0(),
+			},
+		},
+		"k3s_config": {
+			Type:     schema.TypeList,
+			MaxItems: 1,
+			Computed: true,
+			Elem: &schema.Resource{
+				Schema: clusterK3SConfigFields(),
+			},
+		},
+		"eks_config": {
+			Type:     schema.TypeList,
+			MaxItems: 1,
+			Computed: true,
+			Elem: &schema.Resource{
+				Schema: clusterEKSConfigFields(),
+			},
+		},
+		"aks_config": {
+			Type:     schema.TypeList,
+			MaxItems: 1,
+			Computed: true,
+			Elem: &schema.Resource{
+				Schema: clusterAKSConfigFields(),
+			},
+		},
+		"gke_config": {
+			Type:     schema.TypeList,
+			MaxItems: 1,
+			Computed: true,
+			Elem: &schema.Resource{
+				Schema: clusterGKEConfigFields(),
+			},
+		},
+		"default_project_id": {
+			Type:     schema.TypeString,
+			Computed: true,
+		},
+		"system_project_id": {
+			Type:     schema.TypeString,
+			Computed: true,
+		},
+		"description": {
+			Type:     schema.TypeString,
+			Computed: true,
+		},
+		"cluster_auth_endpoint": {
+			Type:     schema.TypeList,
+			MaxItems: 1,
+			Computed: true,
+			Elem: &schema.Resource{
+				Schema: clusterAuthEndpoint(),
+			},
+		},
+		"cluster_monitoring_input": {
+			Type:        schema.TypeList,
+			MaxItems:    1,
+			Computed:    true,
+			Description: "Cluster monitoring configuration",
+			Elem: &schema.Resource{
+				Schema: monitoringInputFields(),
+			},
+		},
+		"cluster_registration_token": {
+			Type:     schema.TypeList,
+			MaxItems: 1,
+			Computed: true,
+			Elem: &schema.Resource{
+				Schema: clusterRegistationTokenFields(),
+			},
+		},
+		"cluster_template_answers": {
+			Type:        schema.TypeList,
+			Computed:    true,
+			MaxItems:    1,
+			Description: "Cluster template answers",
+			Elem: &schema.Resource{
+				Schema: answerFields(),
+			},
+		},
+		"cluster_template_id": {
+			Type:        schema.TypeString,
+			Computed:    true,
+			Description: "Cluster template ID",
+		},
+		"cluster_template_questions": {
+			Type:        schema.TypeList,
+			Computed:    true,
+			Description: "Cluster template questions",
+			Elem: &schema.Resource{
+				Schema: questionFields(),
+			},
+		},
+		"cluster_template_revision_id": {
+			Type:        schema.TypeString,
+			Computed:    true,
+			Description: "Cluster template revision ID",
+		},
+		"default_pod_security_policy_template_id": {
+			Type:        schema.TypeString,
+			Computed:    true,
+			Description: "Default pod security policy template id",
+		},
+		"enable_cluster_alerting": {
+			Type:        schema.TypeBool,
+			Computed:    true,
+			Description: "Enable built-in cluster alerting",
+		},
+		"enable_cluster_monitoring": {
+			Type:        schema.TypeBool,
+			Computed:    true,
+			Description: "Enable built-in cluster monitoring",
+		},
+		"enable_network_policy": {
+			Type:        schema.TypeBool,
+			Computed:    true,
+			Description: "Enable project network isolation",
+		},
+		"scheduled_cluster_scan": {
+			Type:        schema.TypeList,
+			Computed:    true,
+			Description: "Cluster scheduled scan",
+			Elem: &schema.Resource{
+				Schema: scheduledClusterScanFields(),
+			},
+		},
+		"annotations": {
+			Type:     schema.TypeMap,
+			Computed: true,
+		},
+		"labels": {
+			Type:     schema.TypeMap,
+			Computed: true,
+		},
+	}
+	return s
+}
+
+func clusterFieldsV0() map[string]*schema.Schema {
+	s := map[string]*schema.Schema{
+		"name": {
+			Type:     schema.TypeString,
+			Required: true,
+		},
+		"driver": {
+			Type:         schema.TypeString,
+			Optional:     true,
+			Computed:     true,
+			ValidateFunc: validation.StringInSlice(clusterDrivers, true),
+		},
+		"kube_config": {
+			Type:      schema.TypeString,
+			Computed:  true,
+			Sensitive: true,
+		},
+		"rke_config": {
+			Type:          schema.TypeList,
+			MaxItems:      1,
+			Optional:      true,
+			Computed:      true,
+			ConflictsWith: []string{"aks_config", "eks_config", "gke_config", "k3s_config"},
+			Elem: &schema.Resource{
+				Schema: clusterRKEConfigFieldsV0(),
+			},
+		},
+		"k3s_config": {
+			Type:          schema.TypeList,
+			MaxItems:      1,
+			Optional:      true,
+			Computed:      true,
+			ConflictsWith: []string{"aks_config", "eks_config", "gke_config", "rke_config"},
+			Elem: &schema.Resource{
+				Schema: clusterK3SConfigFields(),
+			},
+		},
+		"eks_config": {
+			Type:          schema.TypeList,
+			MaxItems:      1,
+			Optional:      true,
+			ConflictsWith: []string{"aks_config", "gke_config", "k3s_config", "rke_config"},
+			Elem: &schema.Resource{
+				Schema: clusterEKSConfigFields(),
+			},
+		},
+		"aks_config": {
+			Type:          schema.TypeList,
+			MaxItems:      1,
+			Optional:      true,
+			ConflictsWith: []string{"eks_config", "gke_config", "k3s_config", "rke_config"},
+			Elem: &schema.Resource{
+				Schema: clusterAKSConfigFields(),
+			},
+		},
+		"gke_config": {
+			Type:          schema.TypeList,
+			MaxItems:      1,
+			Optional:      true,
+			ConflictsWith: []string{"aks_config", "eks_config", "k3s_config", "rke_config"},
+			Elem: &schema.Resource{
+				Schema: clusterGKEConfigFields(),
+			},
+		},
+		"default_project_id": {
+			Type:     schema.TypeString,
+			Computed: true,
+		},
+		"system_project_id": {
+			Type:     schema.TypeString,
+			Computed: true,
+		},
+		"description": {
+			Type:     schema.TypeString,
+			Optional: true,
+		},
+		"cluster_auth_endpoint": {
+			Type:     schema.TypeList,
+			MaxItems: 1,
+			Optional: true,
+			Computed: true,
+			Elem: &schema.Resource{
+				Schema: clusterAuthEndpoint(),
+			},
+		},
+		"cluster_monitoring_input": {
+			Type:        schema.TypeList,
+			MaxItems:    1,
+			Optional:    true,
+			Description: "Cluster monitoring configuration",
+			Elem: &schema.Resource{
+				Schema: monitoringInputFields(),
+			},
+		},
+		"cluster_registration_token": {
+			Type:     schema.TypeList,
+			MaxItems: 1,
+			Computed: true,
+			Elem: &schema.Resource{
+				Schema: clusterRegistationTokenFields(),
+			},
+		},
+		"cluster_template_answers": {
+			Type:        schema.TypeList,
+			Optional:    true,
+			MaxItems:    1,
+			Computed:    true,
+			Description: "Cluster template answers",
+			Elem: &schema.Resource{
+				Schema: answerFields(),
+			},
+		},
+		"cluster_template_id": {
+			Type:        schema.TypeString,
+			Optional:    true,
+			Description: "Cluster template ID",
+		},
+		"cluster_template_questions": {
+			Type:        schema.TypeList,
+			Optional:    true,
+			Description: "Cluster template questions",
+			Elem: &schema.Resource{
+				Schema: questionFields(),
+			},
+		},
+		"cluster_template_revision_id": {
+			Type:        schema.TypeString,
+			Optional:    true,
+			Description: "Cluster template revision ID",
+		},
+		"default_pod_security_policy_template_id": {
+			Type:        schema.TypeString,
+			Optional:    true,
+			Computed:    true,
+			Description: "Default pod security policy template id",
+		},
+		"desired_agent_image": {
+			Type:     schema.TypeString,
+			Optional: true,
+			Computed: true,
+		},
+		"desired_auth_image": {
+			Type:     schema.TypeString,
+			Optional: true,
+			Computed: true,
+		},
+		"docker_root_dir": {
+			Type:     schema.TypeString,
+			Optional: true,
+			Computed: true,
+		},
+		"enable_cluster_alerting": {
+			Type:        schema.TypeBool,
+			Optional:    true,
+			Computed:    true,
+			Description: "Enable built-in cluster alerting",
+		},
+		"enable_cluster_monitoring": {
+			Type:        schema.TypeBool,
+			Optional:    true,
+			Computed:    true,
+			Description: "Enable built-in cluster monitoring",
+		},
+		"enable_cluster_istio": {
+			Type:        schema.TypeBool,
+			Computed:    true,
+			Description: "Enable built-in cluster istio",
+			Deprecated:  "Deploy istio using rancher2_app resource instead",
+		},
+		"istio_enabled": {
+			Type:        schema.TypeBool,
+			Computed:    true,
+			Description: "Is istio enabled at cluster?",
+		},
+		"enable_network_policy": {
+			Type:        schema.TypeBool,
+			Optional:    true,
+			Computed:    true,
+			Description: "Enable project network isolation",
+		},
+		"scheduled_cluster_scan": {
+			Type:        schema.TypeList,
+			Optional:    true,
+			Computed:    true,
+			MaxItems:    1,
+			Description: "Cluster scheduled scan",
+			Elem: &schema.Resource{
+				Schema: scheduledClusterScanFields(),
+			},
+		},
+		"windows_prefered_cluster": {
+			Type:        schema.TypeBool,
+			Optional:    true,
+			Default:     false,
+			Description: "Windows preferred cluster",
+			ForceNew:    true,
+		},
+	}
+
+	for k, v := range commonAnnotationLabelFields() {
+		s[k] = v
+	}
+
+	return s
+}
+
 func clusterFields() map[string]*schema.Schema {
 	s := map[string]*schema.Schema{
 		"name": {

--- a/rancher2/schema_cluster_rke_config.go
+++ b/rancher2/schema_cluster_rke_config.go
@@ -11,6 +11,183 @@ const (
 
 //Schemas
 
+func clusterRKEConfigFieldsV0() map[string]*schema.Schema {
+	s := map[string]*schema.Schema{
+		"addon_job_timeout": {
+			Type:        schema.TypeInt,
+			Optional:    true,
+			Computed:    true,
+			Description: "Optional duration in seconds of addon job.",
+		},
+		"addons": {
+			Type:        schema.TypeString,
+			Optional:    true,
+			Description: "Optional addons descripton to deploy on rke cluster.",
+			StateFunc:   TrimSpace,
+		},
+		"addons_include": {
+			Type:        schema.TypeList,
+			Optional:    true,
+			Description: "Optional addons yaml manisfest to deploy on rke cluster.",
+			Elem: &schema.Schema{
+				Type: schema.TypeString,
+			},
+		},
+		"authentication": {
+			Type:        schema.TypeList,
+			Description: "Kubernetes cluster authentication",
+			MaxItems:    1,
+			Optional:    true,
+			Computed:    true,
+			Elem: &schema.Resource{
+				Schema: clusterRKEConfigAuthenticationFields(),
+			},
+		},
+		"authorization": {
+			Type:        schema.TypeList,
+			Description: "Kubernetes cluster authorization",
+			MaxItems:    1,
+			Optional:    true,
+			Computed:    true,
+			Elem: &schema.Resource{
+				Schema: clusterRKEConfigAuthorizationFields(),
+			},
+		},
+		"bastion_host": {
+			Type:        schema.TypeList,
+			Description: "RKE bastion host",
+			MaxItems:    1,
+			Optional:    true,
+			Computed:    true,
+			Elem: &schema.Resource{
+				Schema: clusterRKEConfigBastionHostFields(),
+			},
+		},
+		"cloud_provider": {
+			Type:     schema.TypeList,
+			MaxItems: 1,
+			Optional: true,
+			Computed: true,
+			Elem: &schema.Resource{
+				Schema: clusterRKEConfigCloudProviderFields(),
+			},
+		},
+		"dns": {
+			Type:     schema.TypeList,
+			MaxItems: 1,
+			Optional: true,
+			Computed: true,
+			Elem: &schema.Resource{
+				Schema: clusterRKEConfigDNSFields(),
+			},
+		},
+		"ignore_docker_version": {
+			Type:        schema.TypeBool,
+			Optional:    true,
+			Default:     true,
+			Description: "Optional ignore docker version on nodes",
+		},
+		"ingress": {
+			Type:        schema.TypeList,
+			Description: "Kubernetes ingress configuration",
+			MaxItems:    1,
+			Optional:    true,
+			Computed:    true,
+			Elem: &schema.Resource{
+				Schema: clusterRKEConfigIngressFields(),
+			},
+		},
+		"kubernetes_version": {
+			Type:        schema.TypeString,
+			Optional:    true,
+			Computed:    true,
+			Description: "Optional kubernetes version to deploy",
+		},
+		"monitoring": {
+			Type:        schema.TypeList,
+			Description: "Kubernetes cluster monitoring",
+			MaxItems:    1,
+			Optional:    true,
+			Computed:    true,
+			Elem: &schema.Resource{
+				Schema: clusterRKEConfigMonitoringFields(),
+			},
+		},
+		"network": {
+			Type:        schema.TypeList,
+			Description: "Kubernetes cluster networking",
+			MaxItems:    1,
+			Optional:    true,
+			Computed:    true,
+			Elem: &schema.Resource{
+				Schema: clusterRKEConfigNetworkFields(),
+			},
+		},
+		"nodes": {
+			Type:        schema.TypeList,
+			Description: "Optional RKE cluster nodes",
+			Optional:    true,
+			Elem: &schema.Resource{
+				Schema: clusterRKEConfigNodesFields(),
+			},
+		},
+		"prefix_path": {
+			Type:        schema.TypeString,
+			Optional:    true,
+			Computed:    true,
+			Description: "Optional prefix to customize kubernetes path",
+		},
+		"private_registries": {
+			Type:        schema.TypeList,
+			Description: "Optional private registries for docker images",
+			Optional:    true,
+			Elem: &schema.Resource{
+				Schema: clusterRKEConfigPrivateRegistriesFields(),
+			},
+		},
+		"services": {
+			Type:        schema.TypeList,
+			Description: "Kubernetes cluster services",
+			MaxItems:    1,
+			Optional:    true,
+			Computed:    true,
+			Elem: &schema.Resource{
+				Schema: clusterRKEConfigServicesFieldsV0(),
+			},
+		},
+		"ssh_agent_auth": {
+			Type:        schema.TypeBool,
+			Optional:    true,
+			Default:     false,
+			Description: "Optional use ssh agent auth",
+		},
+		"ssh_cert_path": {
+			Type:        schema.TypeString,
+			Optional:    true,
+			Computed:    true,
+			Description: "Optional cluster level SSH certificate path",
+		},
+		"ssh_key_path": {
+			Type:        schema.TypeString,
+			Optional:    true,
+			Computed:    true,
+			Description: "Optional cluster level SSH private key path",
+		},
+		"upgrade_strategy": {
+			Type:        schema.TypeList,
+			Description: "RKE upgrade strategy",
+			MaxItems:    1,
+			Optional:    true,
+			Computed:    true,
+			Elem: &schema.Resource{
+				Schema: clusterRKEConfigNodeUpgradeStrategyFields(),
+			},
+		},
+	}
+
+	return s
+}
+
 func clusterRKEConfigFields() map[string]*schema.Schema {
 	s := map[string]*schema.Schema{
 		"addon_job_timeout": {
@@ -153,6 +330,184 @@ func clusterRKEConfigFields() map[string]*schema.Schema {
 			Computed:    true,
 			Elem: &schema.Resource{
 				Schema: clusterRKEConfigServicesFields(),
+			},
+		},
+		"ssh_agent_auth": {
+			Type:        schema.TypeBool,
+			Optional:    true,
+			Default:     false,
+			Description: "Optional use ssh agent auth",
+		},
+		"ssh_cert_path": {
+			Type:        schema.TypeString,
+			Optional:    true,
+			Computed:    true,
+			Description: "Optional cluster level SSH certificate path",
+		},
+		"ssh_key_path": {
+			Type:        schema.TypeString,
+			Optional:    true,
+			Computed:    true,
+			Description: "Optional cluster level SSH private key path",
+		},
+		"upgrade_strategy": {
+			Type:        schema.TypeList,
+			Description: "RKE upgrade strategy",
+			MaxItems:    1,
+			Optional:    true,
+			Computed:    true,
+			Elem: &schema.Resource{
+				Schema: clusterRKEConfigNodeUpgradeStrategyFields(),
+			},
+		},
+	}
+
+	return s
+}
+
+// Used by datasource
+func clusterRKEConfigFieldsData() map[string]*schema.Schema {
+	s := map[string]*schema.Schema{
+		"addon_job_timeout": {
+			Type:        schema.TypeInt,
+			Optional:    true,
+			Computed:    true,
+			Description: "Optional duration in seconds of addon job.",
+		},
+		"addons": {
+			Type:        schema.TypeString,
+			Optional:    true,
+			Description: "Optional addons descripton to deploy on rke cluster.",
+			StateFunc:   TrimSpace,
+		},
+		"addons_include": {
+			Type:        schema.TypeList,
+			Optional:    true,
+			Description: "Optional addons yaml manisfest to deploy on rke cluster.",
+			Elem: &schema.Schema{
+				Type: schema.TypeString,
+			},
+		},
+		"authentication": {
+			Type:        schema.TypeList,
+			Description: "Kubernetes cluster authentication",
+			MaxItems:    1,
+			Optional:    true,
+			Computed:    true,
+			Elem: &schema.Resource{
+				Schema: clusterRKEConfigAuthenticationFields(),
+			},
+		},
+		"authorization": {
+			Type:        schema.TypeList,
+			Description: "Kubernetes cluster authorization",
+			MaxItems:    1,
+			Optional:    true,
+			Computed:    true,
+			Elem: &schema.Resource{
+				Schema: clusterRKEConfigAuthorizationFields(),
+			},
+		},
+		"bastion_host": {
+			Type:        schema.TypeList,
+			Description: "RKE bastion host",
+			MaxItems:    1,
+			Optional:    true,
+			Computed:    true,
+			Elem: &schema.Resource{
+				Schema: clusterRKEConfigBastionHostFields(),
+			},
+		},
+		"cloud_provider": {
+			Type:     schema.TypeList,
+			MaxItems: 1,
+			Optional: true,
+			Computed: true,
+			Elem: &schema.Resource{
+				Schema: clusterRKEConfigCloudProviderFields(),
+			},
+		},
+		"dns": {
+			Type:     schema.TypeList,
+			MaxItems: 1,
+			Optional: true,
+			Computed: true,
+			Elem: &schema.Resource{
+				Schema: clusterRKEConfigDNSFields(),
+			},
+		},
+		"ignore_docker_version": {
+			Type:        schema.TypeBool,
+			Optional:    true,
+			Default:     true,
+			Description: "Optional ignore docker version on nodes",
+		},
+		"ingress": {
+			Type:        schema.TypeList,
+			Description: "Kubernetes ingress configuration",
+			MaxItems:    1,
+			Optional:    true,
+			Computed:    true,
+			Elem: &schema.Resource{
+				Schema: clusterRKEConfigIngressFields(),
+			},
+		},
+		"kubernetes_version": {
+			Type:        schema.TypeString,
+			Optional:    true,
+			Computed:    true,
+			Description: "Optional kubernetes version to deploy",
+		},
+		"monitoring": {
+			Type:        schema.TypeList,
+			Description: "Kubernetes cluster monitoring",
+			MaxItems:    1,
+			Optional:    true,
+			Computed:    true,
+			Elem: &schema.Resource{
+				Schema: clusterRKEConfigMonitoringFields(),
+			},
+		},
+		"network": {
+			Type:        schema.TypeList,
+			Description: "Kubernetes cluster networking",
+			MaxItems:    1,
+			Optional:    true,
+			Computed:    true,
+			Elem: &schema.Resource{
+				Schema: clusterRKEConfigNetworkFields(),
+			},
+		},
+		"nodes": {
+			Type:        schema.TypeList,
+			Description: "Optional RKE cluster nodes",
+			Optional:    true,
+			Elem: &schema.Resource{
+				Schema: clusterRKEConfigNodesFields(),
+			},
+		},
+		"prefix_path": {
+			Type:        schema.TypeString,
+			Optional:    true,
+			Computed:    true,
+			Description: "Optional prefix to customize kubernetes path",
+		},
+		"private_registries": {
+			Type:        schema.TypeList,
+			Description: "Optional private registries for docker images",
+			Optional:    true,
+			Elem: &schema.Resource{
+				Schema: clusterRKEConfigPrivateRegistriesFields(),
+			},
+		},
+		"services": {
+			Type:        schema.TypeList,
+			Description: "Kubernetes cluster services",
+			MaxItems:    1,
+			Optional:    true,
+			Computed:    true,
+			Elem: &schema.Resource{
+				Schema: clusterRKEConfigServicesFieldsData(),
 			},
 		},
 		"ssh_agent_auth": {

--- a/rancher2/schema_cluster_rke_config_services.go
+++ b/rancher2/schema_cluster_rke_config_services.go
@@ -6,6 +6,66 @@ import (
 
 //Schemas
 
+func clusterRKEConfigServicesFieldsV0() map[string]*schema.Schema {
+	s := map[string]*schema.Schema{
+		"etcd": {
+			Type:     schema.TypeList,
+			MaxItems: 1,
+			Optional: true,
+			Computed: true,
+			Elem: &schema.Resource{
+				Schema: clusterRKEConfigServicesEtcdFields(),
+			},
+		},
+		"kube_api": {
+			Type:     schema.TypeList,
+			MaxItems: 1,
+			Optional: true,
+			Computed: true,
+			Elem: &schema.Resource{
+				Schema: clusterRKEConfigServicesKubeAPIFieldsV0(),
+			},
+		},
+		"kube_controller": {
+			Type:     schema.TypeList,
+			MaxItems: 1,
+			Optional: true,
+			Computed: true,
+			Elem: &schema.Resource{
+				Schema: clusterRKEConfigServicesKubeControllerFields(),
+			},
+		},
+		"kubelet": {
+			Type:     schema.TypeList,
+			MaxItems: 1,
+			Optional: true,
+			Computed: true,
+			Elem: &schema.Resource{
+				Schema: clusterRKEConfigServicesKubeletFields(),
+			},
+		},
+		"kubeproxy": {
+			Type:     schema.TypeList,
+			MaxItems: 1,
+			Optional: true,
+			Computed: true,
+			Elem: &schema.Resource{
+				Schema: clusterRKEConfigServicesKubeproxyFields(),
+			},
+		},
+		"scheduler": {
+			Type:     schema.TypeList,
+			MaxItems: 1,
+			Optional: true,
+			Computed: true,
+			Elem: &schema.Resource{
+				Schema: clusterRKEConfigServicesSchedulerFields(),
+			},
+		},
+	}
+	return s
+}
+
 func clusterRKEConfigServicesFields() map[string]*schema.Schema {
 	s := map[string]*schema.Schema{
 		"etcd": {
@@ -24,6 +84,67 @@ func clusterRKEConfigServicesFields() map[string]*schema.Schema {
 			Computed: true,
 			Elem: &schema.Resource{
 				Schema: clusterRKEConfigServicesKubeAPIFields(),
+			},
+		},
+		"kube_controller": {
+			Type:     schema.TypeList,
+			MaxItems: 1,
+			Optional: true,
+			Computed: true,
+			Elem: &schema.Resource{
+				Schema: clusterRKEConfigServicesKubeControllerFields(),
+			},
+		},
+		"kubelet": {
+			Type:     schema.TypeList,
+			MaxItems: 1,
+			Optional: true,
+			Computed: true,
+			Elem: &schema.Resource{
+				Schema: clusterRKEConfigServicesKubeletFields(),
+			},
+		},
+		"kubeproxy": {
+			Type:     schema.TypeList,
+			MaxItems: 1,
+			Optional: true,
+			Computed: true,
+			Elem: &schema.Resource{
+				Schema: clusterRKEConfigServicesKubeproxyFields(),
+			},
+		},
+		"scheduler": {
+			Type:     schema.TypeList,
+			MaxItems: 1,
+			Optional: true,
+			Computed: true,
+			Elem: &schema.Resource{
+				Schema: clusterRKEConfigServicesSchedulerFields(),
+			},
+		},
+	}
+	return s
+}
+
+// Used by datasource
+func clusterRKEConfigServicesFieldsData() map[string]*schema.Schema {
+	s := map[string]*schema.Schema{
+		"etcd": {
+			Type:     schema.TypeList,
+			MaxItems: 1,
+			Optional: true,
+			Computed: true,
+			Elem: &schema.Resource{
+				Schema: clusterRKEConfigServicesEtcdFields(),
+			},
+		},
+		"kube_api": {
+			Type:     schema.TypeList,
+			MaxItems: 1,
+			Optional: true,
+			Computed: true,
+			Elem: &schema.Resource{
+				Schema: clusterRKEConfigServicesKubeAPIFieldsData(),
 			},
 		},
 		"kube_controller": {

--- a/rancher2/schema_cluster_rke_config_services_kube_api.go
+++ b/rancher2/schema_cluster_rke_config_services_kube_api.go
@@ -123,6 +123,21 @@ func clusterRKEConfigServicesKubeAPIAuditLogFields() map[string]*schema.Schema {
 	return s
 }
 
+func clusterRKEConfigServicesKubeAPIEventRateLimitFieldsV0() map[string]*schema.Schema {
+	s := map[string]*schema.Schema{
+		"configuration": {
+			Type:     schema.TypeMap,
+			Optional: true,
+		},
+		"enabled": {
+			Type:     schema.TypeBool,
+			Optional: true,
+			Default:  false,
+		},
+	}
+	return s
+}
+
 func clusterRKEConfigServicesKubeAPIEventRateLimitFields() map[string]*schema.Schema {
 	s := map[string]*schema.Schema{
 		"configuration": {
@@ -161,6 +176,33 @@ func clusterRKEConfigServicesKubeAPIEventRateLimitFields() map[string]*schema.Sc
 				newMap, _ := ghodssyamlToMapInterface(new)
 				return reflect.DeepEqual(oldMap, newMap)
 			},
+		},
+		"enabled": {
+			Type:     schema.TypeBool,
+			Optional: true,
+			Default:  false,
+		},
+	}
+	return s
+}
+
+// Used by datasource
+func clusterRKEConfigServicesKubeAPIEventRateLimitFieldsData() map[string]*schema.Schema {
+	s := map[string]*schema.Schema{
+		"enabled": {
+			Type:     schema.TypeBool,
+			Optional: true,
+			Default:  false,
+		},
+	}
+	return s
+}
+
+func clusterRKEConfigServicesKubeAPISecretsEncryptionConfigFieldsV0() map[string]*schema.Schema {
+	s := map[string]*schema.Schema{
+		"custom_config": {
+			Type:     schema.TypeMap,
+			Optional: true,
 		},
 		"enabled": {
 			Type:     schema.TypeBool,
@@ -213,6 +255,96 @@ func clusterRKEConfigServicesKubeAPISecretsEncryptionConfigFields() map[string]*
 			Type:     schema.TypeBool,
 			Optional: true,
 			Default:  false,
+		},
+	}
+	return s
+}
+
+// Used by datasource
+func clusterRKEConfigServicesKubeAPISecretsEncryptionConfigFieldsData() map[string]*schema.Schema {
+	s := map[string]*schema.Schema{
+		"enabled": {
+			Type:     schema.TypeBool,
+			Optional: true,
+			Default:  false,
+		},
+	}
+	return s
+}
+
+func clusterRKEConfigServicesKubeAPIFieldsV0() map[string]*schema.Schema {
+	s := map[string]*schema.Schema{
+		"admission_configuration": {
+			Type:     schema.TypeMap,
+			Optional: true,
+		},
+		"always_pull_images": {
+			Type:     schema.TypeBool,
+			Optional: true,
+			Default:  false,
+		},
+		"audit_log": {
+			Type:     schema.TypeList,
+			MaxItems: 1,
+			Optional: true,
+			Elem: &schema.Resource{
+				Schema: clusterRKEConfigServicesKubeAPIAuditLogFields(),
+			},
+		},
+		"event_rate_limit": {
+			Type:     schema.TypeList,
+			MaxItems: 1,
+			Optional: true,
+			Elem: &schema.Resource{
+				Schema: clusterRKEConfigServicesKubeAPIEventRateLimitFieldsV0(),
+			},
+		},
+		"extra_args": {
+			Type:     schema.TypeMap,
+			Optional: true,
+			Computed: true,
+		},
+		"extra_binds": {
+			Type:     schema.TypeList,
+			Optional: true,
+			Elem: &schema.Schema{
+				Type: schema.TypeString,
+			},
+		},
+		"extra_env": {
+			Type:     schema.TypeList,
+			Optional: true,
+			Elem: &schema.Schema{
+				Type: schema.TypeString,
+			},
+		},
+		"image": {
+			Type:     schema.TypeString,
+			Optional: true,
+			Computed: true,
+		},
+		"pod_security_policy": {
+			Type:     schema.TypeBool,
+			Optional: true,
+			Default:  false,
+		},
+		"secrets_encryption_config": {
+			Type:     schema.TypeList,
+			MaxItems: 1,
+			Optional: true,
+			Elem: &schema.Resource{
+				Schema: clusterRKEConfigServicesKubeAPISecretsEncryptionConfigFieldsV0(),
+			},
+		},
+		"service_cluster_ip_range": {
+			Type:     schema.TypeString,
+			Optional: true,
+			Computed: true,
+		},
+		"service_node_port_range": {
+			Type:     schema.TypeString,
+			Optional: true,
+			Computed: true,
 		},
 	}
 	return s
@@ -280,6 +412,85 @@ func clusterRKEConfigServicesKubeAPIFields() map[string]*schema.Schema {
 			Optional: true,
 			Elem: &schema.Resource{
 				Schema: clusterRKEConfigServicesKubeAPISecretsEncryptionConfigFields(),
+			},
+		},
+		"service_cluster_ip_range": {
+			Type:     schema.TypeString,
+			Optional: true,
+			Computed: true,
+		},
+		"service_node_port_range": {
+			Type:     schema.TypeString,
+			Optional: true,
+			Computed: true,
+		},
+	}
+	return s
+}
+
+// Used by datasource
+func clusterRKEConfigServicesKubeAPIFieldsData() map[string]*schema.Schema {
+	s := map[string]*schema.Schema{
+		"admission_configuration": {
+			Type:     schema.TypeMap,
+			Optional: true,
+		},
+		"always_pull_images": {
+			Type:     schema.TypeBool,
+			Optional: true,
+			Default:  false,
+		},
+		"audit_log": {
+			Type:     schema.TypeList,
+			MaxItems: 1,
+			Optional: true,
+			Elem: &schema.Resource{
+				Schema: clusterRKEConfigServicesKubeAPIAuditLogFields(),
+			},
+		},
+		"event_rate_limit": {
+			Type:     schema.TypeList,
+			MaxItems: 1,
+			Optional: true,
+			Elem: &schema.Resource{
+				Schema: clusterRKEConfigServicesKubeAPIEventRateLimitFieldsData(),
+			},
+		},
+		"extra_args": {
+			Type:     schema.TypeMap,
+			Optional: true,
+			Computed: true,
+		},
+		"extra_binds": {
+			Type:     schema.TypeList,
+			Optional: true,
+			Elem: &schema.Schema{
+				Type: schema.TypeString,
+			},
+		},
+		"extra_env": {
+			Type:     schema.TypeList,
+			Optional: true,
+			Elem: &schema.Schema{
+				Type: schema.TypeString,
+			},
+		},
+		"image": {
+			Type:     schema.TypeString,
+			Optional: true,
+			Computed: true,
+		},
+		"pod_security_policy": {
+			Type:     schema.TypeBool,
+			Optional: true,
+			Default:  false,
+		},
+		"secrets_encryption_config": {
+			Type:     schema.TypeList,
+			MaxItems: 1,
+			Optional: true,
+			Elem: &schema.Resource{
+				Schema: clusterRKEConfigServicesKubeAPISecretsEncryptionConfigFieldsData(),
 			},
 		},
 		"service_cluster_ip_range": {

--- a/rancher2/schema_cluster_template.go
+++ b/rancher2/schema_cluster_template.go
@@ -47,6 +47,95 @@ func questionFields() map[string]*schema.Schema {
 	return s
 }
 
+func clusterSpecBaseFieldsV0() map[string]*schema.Schema {
+	s := map[string]*schema.Schema{
+		"cluster_auth_endpoint": {
+			Type:        schema.TypeList,
+			MaxItems:    1,
+			Optional:    true,
+			Computed:    true,
+			Description: "Local cluster auth endpoint",
+			Elem: &schema.Resource{
+				Schema: clusterAuthEndpoint(),
+			},
+		},
+		"default_cluster_role_for_project_members": {
+			Type:        schema.TypeString,
+			Optional:    true,
+			Computed:    true,
+			Description: "Default cluster role for project members",
+		},
+		"default_pod_security_policy_template_id": {
+			Type:        schema.TypeString,
+			Optional:    true,
+			Computed:    true,
+			Description: "Default pod security policy template ID",
+		},
+		"desired_agent_image": {
+			Type:        schema.TypeString,
+			Optional:    true,
+			Computed:    true,
+			Description: "Desired agent image",
+		},
+		"desired_auth_image": {
+			Type:        schema.TypeString,
+			Optional:    true,
+			Computed:    true,
+			Description: "Desired auth image",
+		},
+		"docker_root_dir": {
+			Type:        schema.TypeString,
+			Optional:    true,
+			Computed:    true,
+			Description: "Docker Root Dir",
+		},
+		"enable_cluster_alerting": {
+			Type:        schema.TypeBool,
+			Optional:    true,
+			Default:     false,
+			Description: "Enable built-in cluster alerting",
+		},
+		"enable_cluster_monitoring": {
+			Type:        schema.TypeBool,
+			Optional:    true,
+			Default:     false,
+			Description: "Enable built-in cluster monitoring",
+		},
+		"enable_network_policy": {
+			Type:        schema.TypeBool,
+			Optional:    true,
+			Default:     false,
+			Description: "Enable project network isolation",
+		},
+		"rke_config": {
+			Type:        schema.TypeList,
+			MaxItems:    1,
+			Required:    true,
+			Description: "Rancher Kubernetes Engine Config",
+			Elem: &schema.Resource{
+				Schema: clusterRKEConfigFieldsV0(),
+			},
+		},
+		"scheduled_cluster_scan": {
+			Type:        schema.TypeList,
+			Optional:    true,
+			MaxItems:    1,
+			Description: "Cluster scheduled scan",
+			Elem: &schema.Resource{
+				Schema: scheduledClusterScanFields(),
+			},
+		},
+		"windows_prefered_cluster": {
+			Type:        schema.TypeBool,
+			Optional:    true,
+			Default:     false,
+			Description: "Windows prefered cluster",
+		},
+	}
+
+	return s
+}
+
 func clusterSpecBaseFields() map[string]*schema.Schema {
 	s := map[string]*schema.Schema{
 		"cluster_auth_endpoint": {
@@ -136,6 +225,151 @@ func clusterSpecBaseFields() map[string]*schema.Schema {
 	return s
 }
 
+// Used by datasource
+func clusterSpecBaseFieldsData() map[string]*schema.Schema {
+	s := map[string]*schema.Schema{
+		"cluster_auth_endpoint": {
+			Type:        schema.TypeList,
+			MaxItems:    1,
+			Optional:    true,
+			Computed:    true,
+			Description: "Local cluster auth endpoint",
+			Elem: &schema.Resource{
+				Schema: clusterAuthEndpoint(),
+			},
+		},
+		"default_cluster_role_for_project_members": {
+			Type:        schema.TypeString,
+			Optional:    true,
+			Computed:    true,
+			Description: "Default cluster role for project members",
+		},
+		"default_pod_security_policy_template_id": {
+			Type:        schema.TypeString,
+			Optional:    true,
+			Computed:    true,
+			Description: "Default pod security policy template ID",
+		},
+		"desired_agent_image": {
+			Type:        schema.TypeString,
+			Optional:    true,
+			Computed:    true,
+			Description: "Desired agent image",
+		},
+		"desired_auth_image": {
+			Type:        schema.TypeString,
+			Optional:    true,
+			Computed:    true,
+			Description: "Desired auth image",
+		},
+		"docker_root_dir": {
+			Type:        schema.TypeString,
+			Optional:    true,
+			Computed:    true,
+			Description: "Docker Root Dir",
+		},
+		"enable_cluster_alerting": {
+			Type:        schema.TypeBool,
+			Optional:    true,
+			Default:     false,
+			Description: "Enable built-in cluster alerting",
+		},
+		"enable_cluster_monitoring": {
+			Type:        schema.TypeBool,
+			Optional:    true,
+			Default:     false,
+			Description: "Enable built-in cluster monitoring",
+		},
+		"enable_network_policy": {
+			Type:        schema.TypeBool,
+			Optional:    true,
+			Default:     false,
+			Description: "Enable project network isolation",
+		},
+		"rke_config": {
+			Type:        schema.TypeList,
+			MaxItems:    1,
+			Required:    true,
+			Description: "Rancher Kubernetes Engine Config",
+			Elem: &schema.Resource{
+				Schema: clusterRKEConfigFieldsData(),
+			},
+		},
+		"scheduled_cluster_scan": {
+			Type:        schema.TypeList,
+			Optional:    true,
+			MaxItems:    1,
+			Description: "Cluster scheduled scan",
+			Elem: &schema.Resource{
+				Schema: scheduledClusterScanFields(),
+			},
+		},
+		"windows_prefered_cluster": {
+			Type:        schema.TypeBool,
+			Optional:    true,
+			Default:     false,
+			Description: "Windows prefered cluster",
+		},
+	}
+
+	return s
+}
+
+func clusterTemplateRevisionFieldsV0() map[string]*schema.Schema {
+	s := map[string]*schema.Schema{
+		"cluster_config": {
+			Type:        schema.TypeList,
+			MaxItems:    1,
+			Required:    true,
+			Description: "Cluster configuration",
+			Elem: &schema.Resource{
+				Schema: clusterSpecBaseFieldsV0(),
+			},
+		},
+		"cluster_template_id": {
+			Type:        schema.TypeString,
+			Computed:    true,
+			Description: "Cluster template ID",
+		},
+		"default": {
+			Type:        schema.TypeBool,
+			Optional:    true,
+			Default:     false,
+			Description: "Default cluster template revision",
+		},
+		"enabled": {
+			Type:        schema.TypeBool,
+			Optional:    true,
+			Default:     true,
+			Description: "Enable cluster template revision",
+		},
+		"id": {
+			Type:        schema.TypeString,
+			Computed:    true,
+			Description: "Cluster template revision ID",
+		},
+		"name": {
+			Type:        schema.TypeString,
+			Required:    true,
+			Description: "Cluster template revision name",
+		},
+		"questions": {
+			Type:        schema.TypeList,
+			Optional:    true,
+			Description: "Cluster template questions",
+			Elem: &schema.Resource{
+				Schema: questionFields(),
+			},
+		},
+	}
+
+	for k, v := range commonAnnotationLabelFields() {
+		s[k] = v
+	}
+
+	return s
+}
+
 func clusterTemplateRevisionFields() map[string]*schema.Schema {
 	s := map[string]*schema.Schema{
 		"cluster_config": {
@@ -180,6 +414,105 @@ func clusterTemplateRevisionFields() map[string]*schema.Schema {
 			Description: "Cluster template questions",
 			Elem: &schema.Resource{
 				Schema: questionFields(),
+			},
+		},
+	}
+
+	for k, v := range commonAnnotationLabelFields() {
+		s[k] = v
+	}
+
+	return s
+}
+
+// Used by datasource
+func clusterTemplateRevisionFieldsData() map[string]*schema.Schema {
+	s := map[string]*schema.Schema{
+		"cluster_config": {
+			Type:        schema.TypeList,
+			MaxItems:    1,
+			Required:    true,
+			Description: "Cluster configuration",
+			Elem: &schema.Resource{
+				Schema: clusterSpecBaseFieldsData(),
+			},
+		},
+		"cluster_template_id": {
+			Type:        schema.TypeString,
+			Computed:    true,
+			Description: "Cluster template ID",
+		},
+		"default": {
+			Type:        schema.TypeBool,
+			Optional:    true,
+			Default:     false,
+			Description: "Default cluster template revision",
+		},
+		"enabled": {
+			Type:        schema.TypeBool,
+			Optional:    true,
+			Default:     true,
+			Description: "Enable cluster template revision",
+		},
+		"id": {
+			Type:        schema.TypeString,
+			Computed:    true,
+			Description: "Cluster template revision ID",
+		},
+		"name": {
+			Type:        schema.TypeString,
+			Required:    true,
+			Description: "Cluster template revision name",
+		},
+		"questions": {
+			Type:        schema.TypeList,
+			Optional:    true,
+			Description: "Cluster template questions",
+			Elem: &schema.Resource{
+				Schema: questionFields(),
+			},
+		},
+	}
+
+	for k, v := range commonAnnotationLabelFields() {
+		s[k] = v
+	}
+
+	return s
+}
+
+func clusterTemplateFieldsV0() map[string]*schema.Schema {
+	s := map[string]*schema.Schema{
+		"default_revision_id": {
+			Type:        schema.TypeString,
+			Computed:    true,
+			Description: "Default cluster template revision ID",
+		},
+		"description": {
+			Type:        schema.TypeString,
+			Optional:    true,
+			Description: "Cluster template description",
+		},
+		"members": {
+			Type:        schema.TypeList,
+			Optional:    true,
+			Description: "Cluster template members",
+			Elem: &schema.Resource{
+				Schema: memberFields(),
+			},
+		},
+		"name": {
+			Type:        schema.TypeString,
+			Required:    true,
+			Description: "Cluster template name",
+		},
+		"template_revisions": {
+			Type:        schema.TypeList,
+			Optional:    true,
+			Computed:    true,
+			Description: "Cluster template revisions",
+			Elem: &schema.Resource{
+				Schema: clusterTemplateRevisionFieldsV0(),
 			},
 		},
 	}

